File: apps/web/src/types/schedule.ts
export interface ScheduleTask {
  id: string;
  project_id: string;
  name: string;
  description?: string | null;
  wbs_code?: string | null;
  parent_task_id?: string | null;
  sort_order: number;
  start_date?: string | null;
  end_date?: string | null;
  duration?: number | null;
  baseline_start?: string | null;
  baseline_end?: string | null;
  baseline_duration?: number | null;
  owner_start_date?: string | null;
  owner_end_date?: string | null;
  sub_start_date?: string | null;
  sub_end_date?: string | null;
  percent_complete: number;
  actual_start?: string | null;
  actual_end?: string | null;
  assigned_to?: string | null;
  assigned_to_sub_id?: string | null;
  milestone: boolean;
  is_critical: boolean;
  cost_code_id?: string | null;
  dependencies: TaskDependency[];
  created_by?: string | null;
  created_at: string;
  updated_at: string;
}

export interface TaskDependency {
  id: string;
  predecessor_id: string;
  dependency_type: string;
  lag_days: number;
}

export interface ScheduleDelay {
  id: string;
  project_id: string;
  task_ids: string[];
  delay_days: number;
  reason_category: string;
  responsible_party: string;
  description: string;
  impacts_gc_schedule: boolean;
  impacts_owner_schedule: boolean;
  impacts_sub_schedule: boolean;
  daily_log_id?: string | null;
  rfi_id?: string | null;
  change_order_id?: string | null;
  status: string;
  approved_by?: string | null;
  approved_at?: string | null;
  applied_at?: string | null;
  created_by: string;
  created_at: string;
}

export interface ScheduleVersion {
  id: string;
  project_id: string;
  version_type: string;
  version_number: number;
  title: string;
  notes?: string | null;
  snapshot_data: Record<string, unknown>;
  published_by: string;
  published_at: string;
}

export interface ScheduleConfig {
  id?: string;
  project_id: string;
  schedule_mode: string;
  derivation_method: string;
  owner_buffer_percent?: number | null;
  sub_compress_percent?: number | null;
  health_on_track_max_days: number;
  health_at_risk_max_days: number;
  sub_notify_intervals: number[];
}

export interface ScheduleHealth {
  status: string;
  slippage_days: number;
  on_track_threshold: number;
  at_risk_threshold: number;
}

export interface ScheduleTaskCreateInput {
  name: string;
  description?: string;
  wbs_code?: string;
  parent_task_id?: string;
  start_date?: string;
  end_date?: string;
  duration?: number;
  assigned_to?: string;
  assigned_to_sub_id?: string;
  milestone?: boolean;
  is_critical?: boolean;
}

File: apps/api/app/schemas/schedule.py
"""Pydantic schemas for Schedule (Tasks, Dependencies, Delays, Versions, Config)."""
from datetime import date, datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict


# ============================================================
# TASKS
# ============================================================

class ScheduleTaskCreate(BaseModel):
    name: str
    description: Optional[str] = None
    wbs_code: Optional[str] = None
    parent_task_id: Optional[UUID] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    duration: Optional[int] = None
    assigned_to: Optional[UUID] = None
    assigned_to_sub_id: Optional[UUID] = None
    milestone: bool = False
    is_critical: bool = False
    cost_code_id: Optional[UUID] = None
    sort_order: int = 0


class ScheduleTaskUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    wbs_code: Optional[str] = None
    parent_task_id: Optional[UUID] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    duration: Optional[int] = None
    percent_complete: Optional[int] = None
    assigned_to: Optional[UUID] = None
    assigned_to_sub_id: Optional[UUID] = None
    milestone: Optional[bool] = None
    is_critical: Optional[bool] = None
    cost_code_id: Optional[UUID] = None
    sort_order: Optional[int] = None


class ScheduleTaskResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    project_id: UUID
    name: str
    description: Optional[str] = None
    wbs_code: Optional[str] = None
    parent_task_id: Optional[UUID] = None
    sort_order: int = 0
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    duration: Optional[int] = None
    baseline_start: Optional[date] = None
    baseline_end: Optional[date] = None
    baseline_duration: Optional[int] = None
    owner_start_date: Optional[date] = None
    owner_end_date: Optional[date] = None
    sub_start_date: Optional[date] = None
    sub_end_date: Optional[date] = None
    percent_complete: int = 0
    actual_start: Optional[date] = None
    actual_end: Optional[date] = None
    assigned_to: Optional[UUID] = None
    assigned_to_sub_id: Optional[UUID] = None
    milestone: bool = False
    is_critical: bool = False
    cost_code_id: Optional[UUID] = None
    dependencies: list[dict] = []
    created_by: Optional[UUID] = None
    created_at: datetime
    updated_at: datetime


class ScheduleTaskListResponse(BaseModel):
    data: list[ScheduleTaskResponse]
    meta: dict


# ============================================================
# DEPENDENCIES
# ============================================================

class DependencyCreate(BaseModel):
    predecessor_id: UUID
    successor_id: UUID
    dependency_type: str = "FS"  # FS | SS | FF | SF
    lag_days: int = 0


class DependencyResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    predecessor_id: UUID
    successor_id: UUID
    dependency_type: str
    lag_days: int


# ============================================================
# DELAYS
# ============================================================

class ScheduleDelayCreate(BaseModel):
    task_ids: list[UUID] = []
    delay_days: int
    reason_category: str
    responsible_party: str
    description: str
    daily_log_id: Optional[UUID] = None
    rfi_id: Optional[UUID] = None
    change_order_id: Optional[UUID] = None


class ScheduleDelayResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    project_id: UUID
    task_ids: list = []
    delay_days: int
    reason_category: str
    responsible_party: str
    description: str
    impacts_gc_schedule: bool
    impacts_owner_schedule: bool
    impacts_sub_schedule: bool
    daily_log_id: Optional[UUID] = None
    rfi_id: Optional[UUID] = None
    change_order_id: Optional[UUID] = None
    status: str
    approved_by: Optional[UUID] = None
    approved_at: Optional[datetime] = None
    applied_at: Optional[datetime] = None
    created_by: UUID
    created_at: datetime


class ScheduleDelayListResponse(BaseModel):
    data: list[ScheduleDelayResponse]
    meta: dict


# ============================================================
# VERSIONS
# ============================================================

class SchedulePublishRequest(BaseModel):
    version_type: str = "FULL_SCHEDULE"  # FULL_SCHEDULE | LOOK_AHEAD
    title: Optional[str] = None
    notes: Optional[str] = None


class ScheduleVersionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    project_id: UUID
    version_type: str
    version_number: int
    title: str
    notes: Optional[str] = None
    snapshot_data: dict = {}
    published_by: UUID
    published_at: datetime


class ScheduleVersionListResponse(BaseModel):
    data: list[ScheduleVersionResponse]
    meta: dict


# ============================================================
# CONFIG
# ============================================================

class ScheduleConfigUpdate(BaseModel):
    schedule_mode: Optional[str] = None
    derivation_method: Optional[str] = None
    owner_buffer_percent: Optional[float] = None
    sub_compress_percent: Optional[float] = None
    health_on_track_max_days: Optional[int] = None
    health_at_risk_max_days: Optional[int] = None
    sub_notify_intervals: Optional[list[int]] = None


class ScheduleConfigResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    project_id: UUID
    schedule_mode: str
    derivation_method: str
    owner_buffer_percent: Optional[float] = None
    sub_compress_percent: Optional[float] = None
    health_on_track_max_days: int
    health_at_risk_max_days: int
    sub_notify_intervals: list = []


# ============================================================
# HEALTH
# ============================================================

class ScheduleHealthResponse(BaseModel):
    status: str  # ON_TRACK | AT_RISK | BEHIND
    slippage_days: int
    on_track_threshold: int
    at_risk_threshold: int
